### PR TITLE
[VEN-980][VEN-990][VEN-991] Multiple swap and deflationary token support

### PR DESCRIPTION
**Pull request by defcon022**
* Fixed swap vulnerability. 
* Code modularisation.
* Deflationary tokens support
* updated .gitignore
* Test for deflationary tokens and normal tokens. 

## Description

Resolves #

## Checklist

- [ ] I have updated the documentation to account for the changes in the code.
- [ ] If I added new functionality, I added tests covering it.
- [ ] If I fixed a bug, I added a test preventing this bug from silently reappearing again.
- [ ] My contribution follows [Venus contribution guidelines](docs/CONTRIBUTING.md).

### CODE REVIEW

File: contracts/Swap/SwapRouter.sol [hunk 1]
```diff
@@ -283,26 +606,101 @@ contract SwapRouter is Ownable2StepUpgradeable, ReentrancyGuardUpgradeable, IPan
         }
     }
 
+    // **** SWAP (supporting fee-on-transfer tokens) ****
+    // requires the initial amount to have already been sent to the first pair
+    function _swapSupportingFeeOnTransferTokens(
+        uint256 amountIn,
+        address[] memory path,
+        address _to
+    ) internal virtual returns (uint256[] memory amounts) {
```
Comment (chechu) [inline contracts/Swap/SwapRouter.sol:615]: I would say we are not using the returned value in any place. Could we remove it?

**Review by chechu — COMMENTED** [review]
Would it make sense to extend [PancakeRouter](https://github.com/pancakeswap/pancake-swap-periphery/blob/master/contracts/PancakeRouter.sol), setting WETH to the address of WBNB (as PCS does in the deployed contract [here](https://bscscan.com/address/0x10ed43c718714eb63d5aa57b78b54704e256024e#readContract)), and only add the functions to supply, repay, swapAndSupply and swapAndRepay?

File: contracts/Swap/SwapRouter.sol [hunk 2]
```diff
@@ -283,26 +606,101 @@ contract SwapRouter is Ownable2StepUpgradeable, ReentrancyGuardUpgradeable, IPan
...
+     * @param swapAmount the amount of tokens supply to Venus Market.
+     */
+    function _supply(address path, address vTokenAddress, uint256 swapAmount) internal {
+        TransferHelper.safeApprove(path, vTokenAddress, 0);
+        TransferHelper.safeApprove(path, vTokenAddress, swapAmount);
+        uint256 response = IVToken(vTokenAddress).mintBehalf(msg.sender, swapAmount);
+        if (response != 0) {
+            revert SupplyError(msg.sender, vTokenAddress, response);
+        } else {
+            emit SupplyOnBehalf(msg.sender, vTokenAddress, swapAmount);
```
[the remark this answers is not in the document]
Comment (defcon022) [reply contracts/Swap/SwapRouter.sol]: Fixed. can verify at [commit](https://github.com/VenusProtocol/venus-protocol/pull/204/commits/fab8ab3067ed480ac4c834762a8a1160e9866af8).

File: contracts/Swap/SwapRouter.sol (hunk 1 above)
Comment (defcon022) [reply contracts/Swap/SwapRouter.sol]: Fixed. can verify at [commit](https://github.com/VenusProtocol/venus-protocol/pull/204/commits/fab8ab3067ed480ac4c834762a8a1160e9866af8).

**Comment by defcon022** [comment]
> Would it make sense to extend [PancakeRouter](https://github.com/pancakeswap/pancake-swap-periphery/blob/master/contracts/PancakeRouter.sol), setting WETH to the address of WBNB (as PCS does in the deployed contract [here](https://bscscan.com/address/0x10ed43c718714eb63d5aa57b78b54704e256024e#readContract)), and only add the functions to supply, repay, swapAndSupply and swapAndRepay?

We are customizing the router, the functionality of pairing is upon us and we dont want to give to Pancake router.

**Review by 0xlucian — COMMENTED** [review]
Do we need to verify `path` list contains at least 2 entries?

**Comment by defcon022** [comment]
> Do we need to verify `path` list contains at least 2 entries?

Internally they are checking for that.

**Comment by defcon022** [comment]
> Would it make sense to extend [PancakeRouter](https://github.com/pancakeswap/pancake-swap-periphery/blob/master/contracts/PancakeRouter.sol), setting WETH to the address of WBNB (as PCS does in the deployed contract [here](https://bscscan.com/address/0x10ed43c718714eb63d5aa57b78b54704e256024e#readContract)), and only add the functions to supply, repay, swapAndSupply and swapAndRepay?

PancakeRouter.sol is using =0.6.6 version and we are using 0.8.13 version for new contracts.
And the code is not optimised as we have to refactor the code (use revert, remove safeMath) which we have already done in the swap functionality(Refactored in current implementation). I think we should better keep the code as it is.
Please let me know your thoughts!

**Comment by chechu** [comment]
> PancakeRouter.sol is using =0.6.6 version and we are using 0.8.13 version for new contracts.
> And the code is not optimised as we have to refactor the code (use revert, remove safeMath) which we have already done in the swap functionality(Refactored in current implementation). I think we should better keep the code as it is.
> Please let me know your thoughts!

I see your point. Thanks for the explanation. And what do you think if we split it in two files? one abstract contract in contracts/Swap/lib/PancakeRouter.sol, with our optimised version of the PancakeRouter, and another one in contracts/Swap/SwapRouter.sol, extending the first one, only with our functions. It shouldn't imply an extra gas cost, and we would increase the readability of the code (imho)

**Review by chechu — APPROVED** [review]
LGTM, thanks!

**Pull request merged**